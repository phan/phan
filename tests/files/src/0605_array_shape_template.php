<?php
/**
 * @template TClassName
 * @param array{
 *     class: class-string<TClassName>,
 *     args: array<int, mixed>,
 * } $params
 * @return TClassName
 */
function class_instance_factory(array $params) {
    $class = $params['class'];
    $args = $params['args'];
    return new $class(...$args);
}
$o = class_instance_factory(['class' => stdClass::class, 'args' => []]);
echo $o->method();

/**
 * @template T1
 * @template T2
 * @param array{0:T1, 1:T2} $params
 * @return array{0:T2, 1:T1}
 */
function swap_pair(array $params) : array {
    return [$params[1], $params[0]];
}
echo strlen(swap_pair([rand(0,5), new stdClass()]));
