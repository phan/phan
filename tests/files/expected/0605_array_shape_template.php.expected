%s:16 PhanUndeclaredMethod Call to undeclared method \stdClass::method
%s:27 PhanTypeMismatchArgumentInternal%SReal Argument 1 ($string) is swap_pair([rand(0, 5),new stdClass()]) of type array{0:\stdClass,1:int} (real type array) but \strlen() takes string
